contracts/adapters/swap/osmosis-poolmanager/src/contract.rs

```diff
@@ -274,3 +291,47 @@ fn query_simulate_swap_exact_asset_out(
     }
     .into())
 }
+
+// Queries the osmosis poolmanager module to simulate a swap exact amount in with spot price
+fn query_simulate_swap_exact_asset_in_with_spot_price(
```

Comment:
This is doing what I had in mine for astroport, but I guess the contract extends to it better.

Reply:
Yeah for Astroport we get the spot price from the simulation responses themselves, so had initially coupled the logic together so that it had one iteration. Now with the new design of getting the responses during simulation and then iterating in the future for spot price, this gets a tad cleaner, but not as clean as the decoupled osmosis version since they actually have spot price queries.